README.md
# devcontainers
Devcontainers or Codespace with optimizations for PHP and Nodejs development.

[![buildx-8.1-16-amd64](https://github.com/madalinignisca/devcontainers/actions/workflows/buildx-bake-8.1-16-amd64.yml/badge.svg)](https://github.com/madalinignisca/devcontainers/actions/workflows/buildx-bake-8.1-16-amd64.yml) [![buildx-8.0](https://github.com/madalinignisca/devcontainers/actions/workflows/buildx-bake-8.0.yml/badge.svg)](https://github.com/madalinignisca/devcontainers/actions/workflows/buildx-bake-8.0.yml) [![buildx-7.4](https://github.com/madalinignisca/devcontainers/actions/workflows/buildx-bake-7.4.yml/badge.svg)](https://github.com/madalinignisca/devcontainers/actions/workflows/buildx-bake-7.4.yml)

Laravel, Symfony, CakePHP, Codeigniter, AdonisJS, NestJS, SailsJS, WordPress, Drupal, Magento,
Prestashop, Opencart compatible or any NodeJS and PHP project.

## Getting started:

Using a shell compatible prompt, download generate.sh and run it.

PHP and Node combinations:
|    | 7.4 | 8.0 | 8.1 |
|----|-----|-----|-----|
| 14 | X   | X   |     |
| 16 |     | X   | X   |
| 18 |     |     | X   |

ARM64 users, you might need to look for supported docker images alternatives for some services. Some official providers don't look into supporting ARM64 too soon.
More on ARM64 soon.

Continue with the [Wiki](https://github.com/madalinignisca/devcontainers/wiki) for advanced documentation.

## Important for Windows 10 users:

https://docs.microsoft.com/en-us/windows/wsl/wsl-config#configure-global-options-with-wslconfig

Make sure you will set a memory limit, as the 80% default limit will allow WSL2 to do aggresive caching in ram
and will simply make your system slow, sometimes even freezing minutes. This is not a bug of Docker, neither of
Linux. It's purely a missed optimization of Microsoft in WSL2, forgetting that Linux, like Windows as well, is
very aggressive on caching anything possible.

Example: I use 8GB memory limit and 0 swap to make docker behave identical like on a virtual server with 8gb of ram
and 2 cpus. Works fine on a 16GB laptop, no slowness while running devcontainers with lots of services, including
Elastic Search. On a 8GB Windows Host, I would set 4GB memory limit, and tweak in `docker-compose.yaml` to enforce
lower memory limits per services, but do really investigate if those services can run with extrem low memory constrains.


CHANGELOG
Sat Sep  3 09:30:46 CEST 2022

Grouped back arm and amd for 8.1 images.


generate.sh
#!/usr/bin/env bash

if [[ $1 = selfupdate ]]; then
  echo "I will update myself"
  cp $(realpath "$0") $(realpath "$0").bak
  curl -o $(realpath "$0") https://raw.githubusercontent.com/madalinignisca/devcontainers/master/generate.sh
  chmod +x $(realpath "$0")
  exit 0
fi

echo "Project name (must be a-z0-9):"
read PRJNAME

if [[ -z $PRJNAME ]]; then
  echo "Must use a project name!"
  exit 1
fi

mkdir $PRJNAME
cd $PRJNAME

echo "services:

" > docker-compose.yaml

echo "Pick a PHP-NodeJS combination":
echo "[1]: PHP 8.1 / NodeJS 18"
echo "[2]: PHP 8.1 / NodeJS 16 (default)"
echo "[3]: PHP 8.0 / NodeJS 16"
echo "[4]: PHP 8.0 / NodeJS 14"
echo "[5]: PHP 7.4 / NodeJS 14"
echo "[6]: PHP 7.4 / NodeJS 12"
read VERSION

case $VERSION in

  1)
    PHPNODEJS="8.1-18"
    ;;

  2)
    PHPNODEJS="8.1-16"
    ;;

  3)
    PHPNODEJS="8.0-16"
    ;;

  4)
    PHPNODEJS="8.0-14"
    ;;

  5)
    PHPNODEJS="7.4-14"
    ;;

  6)
    PHPNODEJS="7.4-12"
    ;;

  *)
    PHPNODEJS="8.1-16"
    ;;

esac

echo "  dev:
    image: 16nsk/devcontainers:$PHPNODEJS
    command: sleep infinity
    volumes:
      - projects:/projects
    ports:
      - \"${BACKEND_PORT:-8000}:8000\"
      - \"${FRONTEND_PORT:-3000}:3000\"
" >> docker-compose.yaml

echo "POSTGRESQL:"
echo "[1] Postgresql 14"
echo "[2] Postgresql 14 with Postgis 3.2"
echo "[3] Postgresql 12 with Postgis 2.5 (for old projects)"
echo "[ ] Hit enter to skip Postgresql"
read POSTGRESQL

case $POSTGRESQL in

  1)
    POSTGRESQL="postgres:14"
    ;;

  2)
    POSTGRESQL="postgis/postgis:14-3.2"
    ;;

  3)
    POSTGRESQL="postgis/postgis:12-2.5"
    ;;

esac

if [[ -n $POSTGRESQL ]]; then
echo "  pgsql:
    image: $POSTGRESQL
    environment:
      - POSTGRES_PASSWORD=developer
      - POSTGRES_USER=developer
      - POSTGRES_DB=developer
      - PGDATA=/var/lib/postgresql/data/pgdata
    volumes:
      - postgres:/var/lib/postgresql/data
    ports:
      - \"${FORWARD_DB_PORT:-5432}:5432\"
" >> docker-compose.yaml
fi


echo "MARIADB:"
echo "[1] MariaDB 10.7"
echo "[2] MariaDB 10.6 (LTS)"
echo "[3] MariaDB 10.3 (Oldest supported)"
echo "[ ] Hit enter to skip MariaDB"
read MARIADB

case $MARIADB in

  1)
    MARIADB="10.7"
    ;;

  2)
    MARIADB="10.6"
    ;;

  3)
    MARIADB="10.3"
    ;;

esac

if [[ -n $MARIADB ]]; then
echo "  mariadb:
    image: mariadb:$MARIADB
    environment:
      - MARIADB_ROOT_PASSWORD=root
      - MARIADB_DATABASE=developer
      - MARIADB_USER=developer
      - MARIADB_PASSWORD=developer
      - MARIADB_AUTO_UPGRADE=true
      - MARIADB_DISABLE_UPGRADE_BACKUP=true
    volumes:
      - mariadb:/var/lib/mysql
    ports:
      - \"${FORWARD_DB_PORT:-3306}:3306\"
" >> docker-compose.yaml
fi

if [[ -n $POSTGRESQL || -n $MARIADB ]]; then
echo "  adminer:
    image: adminer
    ports:
      - \"${FORWARD_ADMINER_PORT:-8080}:8080\"
" >> docker-compose.yaml
fi

echo "Want mailhog?"
read MAILHOG
if [[ -n $MAILHOG ]]; then
echo "  mailhog:
    image: 16nsk/mailhog
    ports:
      - \"${FORWARD_MAILHOG_PORT:-8025}:8025\"
" >> docker-compose.yaml
fi

echo "Want memcached?"
read MEMCACHED
if [[ -n $MEMCACHED ]]; then
echo "  memcached:
    image: memcached:1.6-alpine
" >> docker-compose.yaml
fi

echo "Want redis?"
read REDIS
if [[ -n $REDIS ]]; then
echo "  redis:
    image: redis:7.0-alpine
    command: redis-server --save 60 1 --loglevel warning
    volumes:
      - redis:/data
" >> docker-compose.yaml
fi

echo "volumes:
  projects:" >> docker-compose.yaml

if [[ -n $POSTGRESQL ]]; then
echo "  postgres:" >> docker-compose.yaml
fi

if [[ -n $MARIADB ]]; then
echo "  mariadb:" >> docker-compose.yaml
fi

if [[ -n $REDIS ]]; then
echo "  redis:" >> docker-compose.yaml
fi

echo "{
    \"name\": \"$PRJNAME\",
    \"dockerComposeFile\": \"docker-compose.yaml\",
    \"service\": \"dev\",
    \"workspaceFolder\": \"/projects/workspace\",
    \"shutdownAction\": \"stopCompose\",
    \"settings\": {
        \"terminal.integrated.defaultProfile.linux\": \"bash (login)\",
        \"terminal.integrated.profiles.linux\": {
            \"bash (login)\": {
                \"path\": \"bash\",
                \"args\": [\"-l\"],
                \"icon\": \"terminal-ubuntu\"
            }
        }
    },
    \"extensions\": [
        \"EditorConfig.EditorConfig\",
        \"bmewburn.vscode-intelephense-client\",
        \"calebporzio.better-phpunit\",
        \"dbaeumer.vscode-eslint\",
        \"eamodio.gitlens\",
        \"octref.vetur\",
        \"xdebug.php-debug\"
    ]
}" > .devcontainer.json

if hash docker-compose 2>/dev/null; then
        docker-compose pull
    else
        docker compose pull
fi

echo "All done! Open the folder in Visual Studio Code and run in Containers."
